feat: Implementing Admin 2FA/Dashboard api's, Session ID api/cookies and Middleware

Zyeoh883 opened the pull request:
Implemented the api calls in admin_2FA and Admin_dashboard. Created a sessionID api that stores in cookies. Middleware routes admin and users for /login, /admin_2FA and /admin_dashboard based on the SessionID alone.

### SessionID
- Value is stored as a https cookie
- Is not api protected, email validation is next

### Middleware
- Protect the Dashboard api's
- Routes Admin with valid SessionID from '/' and '/admin_2FA' to dashboard page.
- Routes Non-valid SessionID to '/', does not delete the invalid SessionID (Not tested)

Zyeoh883 committed:
Merge pull request #10 from 42KL-admin/dev

Updating admin_dashboard, working on api implementation for admin dashboard

Zyeoh883 committed:
Made login page route to admin_2FA is it is an admin

Zyeoh883 committed:
Made api's to GET, POST, PUT and DELETE users from mongoDB. They are used by the admin_dashboard and the POST and GET is implememnted. Will implement the other 2 next.

Zyeoh883 committed:
Chnaging the role of the user is now working but there is a clash in naming convention between the store and api's with permission and role respectively. Will now work on the DELTE implementation.

Zyeoh883 committed:
Deleting a user is now working. A general update to all api calls and store updates, an api call is done first and only if successful does it only change the store and update on the front end. A problem right now with the Dashboard is that it does not do live updates only when refreshed

Zyeoh883 committed:
touchup with the emailrow component

Zyeoh883 committed:
Changed EmailPermission t0 EmailRole

Ricky0625 committed:
chore: untrack 42_RC.code-workspace

Ricky0625 committed:
chore: added prettierrc.json

Ricky0625 committed:
feat: add js-cookie

Ricky0625 committed:
feat: add cookie and loggedUserStore

Ricky0625 committed:
feat: cookie for user is done

Ricky0625 committed:
chore: display logged user's email, conditionally display button to
admin dashboard

Ricky0625 committed:
fix: nested html and body in (Admin) route

Ricky0625 committed:
chore: move (admin) route group under (authenticated)

Ricky0625 committed:
chore: change (admin) route to protect from invalid access

Ricky0625 committed:
chore: protect admin_2FA from unauthorized access

Ricky0625 committed:
fix: add option to dropdown menu for admin to access to upload page

Zyeoh883 committed:
Login routes the user to the approriate pages. 2FA Page sends an OTP to the user's email upon entering the webpage, a resend button with a cooldown of 5 seconds is also made. There's some funny interaction with the cookies where there are force routes on certain webpages that will be fixed later.

Zyeoh883 committed:
bug: added middleware but now 2FA does not enter admin_dashboard

OTP_verification is supposed to be called in 2FA page and it also deletes previous OTPs if successful. Unit testing using Postman shows OTP_verification works but implementation is not working as the API is never called or never successful? Once a valid OTP is given on the 2FA page, user is routed to login page for some reason.

Ricky0625 committed:
fix: updated admin_2FA

- add onSubmit event on form, and trigger handleSubmit
- call the correct api: createSessionID -> SessionID
- call setEmail

Ricky0625 committed:
fix: update SessionID route

Disable secure:true. This should be enabled in production only.

Ricky0625 committed:
chore: removed console.log

Ricky0625 committed:
chore: new package: cookies

Zyeoh883 committed:
fix: Middleware is now working to redirect away from admin_dashboard

Had to add the following code into the middleware call for the SessionID GET
request:
'Cookie': `sessionID=${sessionID}`

Also removed the messages for check if sessionExists in SessionID GET to
get it working without prompting the messages. Will add back later.

Zyeoh883 committed:
feat: middleware now handles routes better

Not fully completed but it now reroutes a valid admin from login and 2FA
to dashabord.
API calls (only /admin/Dashboard) are protected and routes
user to login page.

Zyeoh883 committed:
feat: Logout now clears SessionID from cookies and reroutes to '/'

Added DELETE api for cookies to delete locally and on MongoDB.
Edited DropdownMenu.tsx to call DELETE api in /api/admin/SessionID.

Zyeoh883 committed:
bugs: 'useReducer' Error

Archiving this commit to check previous commits.
Attempting to simplify middleware by calling a validateSession function
rather than api call. Added the function as /lib/validateSession.ts.

Zyeoh883 committed:
feat: improvements in middleware redirects

Abstracted middleware admin pages logic to /lib/middleware/validateSession.ts.
Remove routing in src/app/(authenticated)/layout.tsx.
Added router.push to '/' in admin_dashboard page and @hero.

Zyeoh883 committed:
fix: Fixed where EmailOtp is not sent

Middleware prevented access to EmailOtp, need to check the api call if
it was called at Admin_2FA

Zyeoh883 committed:
fix: Fixed run build errors

timeoutID in 2FA_page is adjusted to prevent type errors.
Added router in the dependency array in 2FA layout.
Changed back tick in Admin_dashboard

Zyeoh883 committed:
fix: DropdownMenu now logouts correctly

logoutUser gets the status from the response directly.

Zyeoh883 committed:
Merge branch 'dev' into implementation

Zyeoh883 committed:
chore: updated package-lock.json

Zyeoh883 committed:
fix: fixed dependencies for build

Ricky0625 commented on .prettierrc.json:
Is it possible to use 2 spaces instead of 2 tabs? I think 2 tabs for indentation is too big lol

Ricky0625 replied:
@Zyeoh883

Zyeoh883 committed:
fix: EmailOtp & OtpVerification is fixed

Both EmailOtp and OTPVerification api is now working, bcrypt.compare
should not be used as == is sufficient.

Attempting to reroute admin to home page if sessionID is invalid when
using the components in the dashbaord page. It console.logs that it is
in home but it remains in Daashboard until the refresh?
At least it prevents invalid SessionID from touching anything

Invalid SessionID will prompt the middleware to delete all cookies and
attempt to route to home.

Zyeoh883 committed:
chore: Switched from 2 tabs to 2 spaces

.prettierrc.json is updated to have 2 spaces rather than 2 tabs for
width.

Zyeoh883 committed:
chore: updated package-lock.json

Ricky0625 commented:
**Issue:**
Normal users seem to have issues logging out, while admins can log out without an issue. Upon inspection, the `logoutUser` is trying to call a DELETE API to delete a session ID. This is not valid for normal users, as they don't have any session IDs in the current implementation.

**Suggestion:**
Check if the user's role is admin. If that is true, call DELETE session ID and delete the user cookie. Otherwise, just delete the user cookie in the browser.

Zyeoh883 committed:
fix: Fix DropdownMenu Component not loging out for users

The logoutUser calls a DELETE api to delete SessionID even for users. It
fails then an error is thrown before it can clear and redirect.
Made it where it deletes the email cookie first, calls the DELETE api if
admin and then redirects to home.

Ricky0625 reviewed (approved):
- Tested middleware for admin routes. It seems to be working well.
- Login, 2FA, and logout work fine for admins.
- Login and logout work fine for normal users.
- The user's cookie was stored properly in the browser.
- The user's cookies are cleared after logging out.
